src/ui/ActionContainer.tsx

```diff
@@ -273,49 +274,92 @@ export const ActionContainer = ({
   );
 
   const [executionState, dispatch] = useReducer(executionReducer, {
-    status:
-      overallState !== 'malicious' && isPassingSecurityCheck
-        ? 'idle'
-        : 'blocked',
+    status: 'checking-supportability',
   });
 
-  // in case, where action or websiteUrl changes, we need to reset the action state
+  // in case, where initialAction or websiteUrl changes, we need to reset the action state
   useEffect(() => {
     if (action === initialAction || action.isChained) {
       return;
     }
 
     setAction(initialAction);
     setActionState(getOverallActionState(initialAction, websiteUrl));
-    dispatch({ type: ExecutionType.RESET });
-  }, [action, initialAction, websiteUrl]);
+    dispatch({ type: ExecutionType.CHECK_SUPPORTABILITY });
+    // we want to run this one when initialAction or websiteUrl changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [initialAction, websiteUrl]);
 
   useEffect(() => {
     callbacks?.onActionMount?.(
       action,
       websiteUrl ?? action.url,
       actionState.action,
     );
-    // we ignore changes to `actionState.action` explicitly, since we want this to run once
+    // we ignore changes to `actionState.action` or callbacks explicitly, since we want this to run once
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [callbacks, action, websiteUrl]);
+  }, [action, websiteUrl]);
+
+  useEffect(() => {
+    const dynamicDataConfig = action.dynamicData_experimental;
+    if (
+      !dynamicDataConfig ||
+      !dynamicDataConfig.enabled ||
+      executionState.status !== 'idle' ||
+      action.isChained
+    ) {
+      return;
+    }
+
+    let timeout: any; // NodeJS.Timeout
+    const fetcher = async () => {
+      try {
+        const newAction = await action.refresh();
+
+        // if after refresh user clicked started execution, we should not update the action
+        if (executionState.status === 'idle') {
+          setAction(newAction);
+        }
+      } catch (e) {
+        console.error(
+          `[@dialectlabs/blinks] Failed to fetch dynamic data for action ${action.url}`,
+        );
+        // if fetch failed, we retry after the same delay
+        timeout = setTimeout(fetcher, dynamicDataConfig.delayMs);
```

Comment:
suggestion - wdyt about setting a soft limit on max update freq to avoid possible abuse? e.g. 1 sec

Reply:
Good call! Let's do min 1s

Reply:
Adjusted in `Action` class